modules/photo/src/ccm/ccm.cpp

```diff
@@ -188,20 +194,29 @@ void ColorCorrectionModel::Impl::calWeightsMasks(const Mat& weightsList_, double
     maskedLen = (int)sum(mask)[0];
 }
 
-void ColorCorrectionModel::Impl::initialWhiteBalance(void)
+void ColorCorrectionModel::Impl::initialWhiteBalance()
 {
-    Mat schannels[4];
-    split(srcRgbl, schannels);
-    Mat dchannels[4];
-    split(dstRgbl, dchannels);
-    std::vector<double> initialVec = { sum(dchannels[0])[0] / sum(schannels[0])[0], 0, 0, 0,
-        sum(dchannels[1])[0] / sum(schannels[1])[0], 0, 0, 0,
-        sum(dchannels[2])[0] / sum(schannels[2])[0], 0, 0, 0 };
-    std::vector<double> initialVec_(initialVec.begin(), initialVec.begin() + shape);
-    Mat initialWhiteBalanceMat = Mat(initialVec_, true).reshape(0, shape / 3);
-    ccm0 = initialWhiteBalanceMat;
+    // sum over all pixels – Scalar holds per-channel sums
+    const cv::Scalar srcSum = cv::sum(srcRgbl);
+    const cv::Scalar dstSum = cv::sum(dstRgbl);
+
+    // channel-wise gain factors
+    const double gR = dstSum[0] / srcSum[0];
+    const double gG = dstSum[1] / srcSum[1];
+    const double gB = dstSum[2] / srcSum[2];
+
+    // diagonal CCM with those gains (extra zeros kept for CCM_AFFINE case)
+    std::vector<double> coeffs = {
+        gR, 0, 0, 0,
+        gG, 0, 0, 0,
+        gB, 0, 0, 0
+    };
+    coeffs.resize(shape);                       // shape = 9 (linear) or 12 (affine)
+
+    ccm0 = cv::Mat(coeffs, true).reshape(0, shape / 3);   // 3×3 or 3×4
```

Comment:
The code with ccm0 and coeffs looks very cryptic. I proposr to create ccm0 with proper shape with zeros , e.g. cv::Mat::zeros() and then set non-zero values for it. It's less amount of code, but there is no strange resize/reshape and `shape/3` values.